pkg/networkservice/common/monitor/client.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package monitor
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/heal"
+	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
+)
+
+type monitorClient struct {
+	grpcClient    networkservice.MonitorConnectionClient
+	healer        heal.Healer
```

Comment:
Why monitor client should depend on `heal`l?

Reply:
It should somehow initiate the healing process when connection breaks down. Another way i see is to pass healing functions from `healClient` within context of connection request.

Reply:
Can we just use dependency injection?
For example, we can create an interface and pass it into `monitor.Client`
``` go
type ConnectionEventHandler interface {
    Handle(*ConnectionEvent)
}
```
And for `heal` case we need to just implement this interface. 
For example:

```go
type healConnectionEventHandler struct {
}
func (h *healConnectionEventHandler) Handle(e *ConnectionEvent) {
// StartHeal
}
func NewHealConnectionEventHandler() ConnectionEventHandler {
    return &healConnectionEventHandler{}
}
```

Reply:
As i understood this way we don't need `monitoring` and `reported` maps in `monitor.Client` -- all this job will be done by `heal.Client`, so we get more concrete responsibilities of each client, cool! We also need to add another method to `healConnectionEventHandler` that triggers when connection to `MonitorConnectionServer` breaks down itself. I'll try to implement your suggestion.

Reply:
What is the underlying motivation of this split?

Reply:
nice catch!
@lvfxx Let's create an Issue for this PR.

>What is the underlying motivation of this split?

In short, we want to extract monitoring client from heal pkg to reuse it in different places. For example for `dnscontext.Client` component.

Reply:
@denis-tingajkin But why... the amount of reuse is small... but it introduces relatively complex new relationships... I don't mind replicating code in places if it's small and makes it easier to comprehend in one place what's happening.